learning_paths/api/v1/views.py

```diff
@@ -105,3 +118,233 @@ def get(self, request, learning_path_uuid):
+            status=status.HTTP_204_NO_CONTENT,
+        )
+
+
+class FetchEnrollmentsView(APIView):
+    """
+    A view to fetch learning path enrollments for the authenticated user.
+    """
+
+    permission_classes = [IsAuthenticated]
+
+    def get(self, request):
+        """
+        Handle GET requests to fetch enrollments.
+        """
+        if request.user.is_staff or request.user.is_superuser:
+            enrollments = LearningPathEnrollment.objects.all()
+        else:
+            enrollments = LearningPathEnrollment.objects.filter(user=request.user)
+
+        return Response(LearningPathEnrollmentSerializer(enrollments, many=True).data)
+
+
+class BulkEnrollAPIView(APIView):
+    """
+    Bulk enrollment API for LearningPathEnrollment.
+
+    For existing users, it creates a new LearningPathEnrollment record.
+    For non-existing users, it creates a new LearningPathEnrollmentAllowed record.
+    """
+
+    permission_classes = [IsAdminUser]
+
+    def post(self, request, *args, **kwargs):
+        """
+        POST requests handler for the view.
+        """
+        data = request.data
+        learning_paths_keys = data.get("learning_paths", "").split(",")
+        emails = data.get("emails", "").split(",")
+
+        valid_learning_paths_keys = []
+        for key in learning_paths_keys:
+            try:
+                UUID(key)
+            except ValueError:
+                logger.warning("BulkEnrollAPIView: Invalid learning path key: %s", key)
+                continue
+            valid_learning_paths_keys.append(key)
+
+        learning_paths = LearningPath.objects.filter(
+            uuid__in=valid_learning_paths_keys
+        ).all()
+
+        existing_users = User.objects.filter(email__in=emails).all()
+        non_existing_emails = set(emails) - set(u.email for u in existing_users)
+
+        enrollments_created = []
+        enrollment_allowed_created = []
+
+        for learning_path in learning_paths:
+
+            # Create LearningPathEnrollment, LearningPathEnrollmentAllowed for existing users
+            for user in existing_users:
+                enrollment = LearningPathEnrollment.objects.filter(
+                    user=user, learning_path=learning_path
+                ).first()
+                enrolled_now = False
+                if not enrollment:
+                    enrollment = LearningPathEnrollment(
+                        user=user,
+                        learning_path=learning_path,
+                        is_active=True,
+                        enrolled_at=datetime.now(timezone.utc),
+                    )
+                    enrolled_now = True
+                if not enrollment.is_active:
+                    enrollment.is_active = True
+                    enrollment.enrolled_at = datetime.now(timezone.utc)
+                    enrolled_now = True
+                enrollment.save()
+                if enrolled_now:
+                    enrollments_created.append(enrollment)
```

Comment:
Nit: why don't we use [update_or_create](https://docs.djangoproject.com/en/4.2/ref/models/querysets/#update-or-create) for all of this?

Reply:
@Agrendalath Good question. There is an edge case. When there is an active enrollment for the learning `update_or_create` will change the "enrolled_at" date of the enrollment. I couldn't find a way to set `is_active` for inactive enrollments, but not change `enrolled_at` for active enrollments.